File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/promotion/LocalPromotionSource.java
package com.heaton.funnyvote.data.promotion;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import com.heaton.funnyvote.data.local.dao.PromotionDao;
import com.heaton.funnyvote.database.Promotion;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.utils.AppExecutors;

import java.util.List;

import io.reactivex.Observable;

public class LocalPromotionSource implements PromotionDataSource {
    private PromotionDao promotionDao;
    private static volatile LocalPromotionSource INSTANCE;
    private AppExecutors mAppExecutors;

    private LocalPromotionSource(@NonNull PromotionDao promotionDao, AppExecutors appExecutors) {
        this.promotionDao = promotionDao;
        this.mAppExecutors = appExecutors;
    }

    public static LocalPromotionSource getInstance(@NonNull PromotionDao promotionDao
            , AppExecutors appExecutors) {
        if (INSTANCE == null) {
            synchronized (LocalPromotionSource.class) {
                if (INSTANCE == null) {
                    INSTANCE = new LocalPromotionSource(promotionDao, appExecutors);
                }
            }
        }
        return INSTANCE;
    }

    @Override
    public Observable<List<Promotion>> getPromotionList(User user) {
        return Observable.fromCallable(() -> promotionDao.loadAll());
    }

    @Override
    public void savePromotionList(final List<Promotion> promotionList) {
        mAppExecutors.diskIO().execute(() -> promotionDao.insertOrReplaceInTx(promotionList));
    }

    @VisibleForTesting
    public static void clearInstance() {
        INSTANCE = null;
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/promotion/LocalPromotionSource.kt
package com.heaton.funnyvote.data.promotion

import android.support.annotation.VisibleForTesting

import com.heaton.funnyvote.database.Promotion
import com.heaton.funnyvote.database.PromotionDao
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.utils.AppExecutors

class LocalPromotionSource
private constructor(
        private val promotionDao: PromotionDao,
        private val mAppExecutors: AppExecutors
) : PromotionDataSource {

    override fun getPromotionList(user: User, callback: PromotionDataSource.GetPromotionsCallback) {
        val runnable = Runnable {
            if (user.userCode.isNullOrBlank()) {
                callback.onPromotionsNotAvailable()
            }
            mAppExecutors.mainThread.execute {
                val list = promotionDao.loadAll()
                callback.onPromotionsLoaded(list)
            }
        }
        mAppExecutors.diskIO.execute(runnable)
    }

    override fun savePromotionList(promotionList: List<Promotion>) {
        mAppExecutors.diskIO.execute { promotionDao.insertOrReplaceInTx(promotionList) }
    }

    companion object {
        private var INSTANCE: LocalPromotionSource? = null

        fun getInstance(promotionDao: PromotionDao, appExecutors: AppExecutors): LocalPromotionSource? {
            if (INSTANCE == null) {
                synchronized(LocalPromotionSource::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = LocalPromotionSource(promotionDao, appExecutors)
                    }
                }
            }
            return INSTANCE
        }

        @VisibleForTesting
        fun clearInstance() {
            INSTANCE = null
        }
    }
}
